src/rsz/src/RepairSetup.cc

```diff
@@ -885,76 +886,117 @@ RepairSetup::fanout(Vertex *vertex)
-          return;
-        }
-        status = Operator::op_one;
-    }
-    else if (status == curr_op) {
-        // handle > 2 input scenarios (up to any arbitrary number)
-        getEquivPortList2(expr->left(), ports, inv_ports, status);
-        if (status == Operator::op_port) {
-            return;
-        }
-        getEquivPortList2(expr->right(), ports, inv_ports, status);
-        if (status == Operator::op_port) {
-            return;
-        }
-    }
-    else if (curr_op == Operator::op_port && expr->port() != nullptr) {
-        ports.insert(expr->port());
-    }
-    else {
-        status = Operator::op_port; // moved to some other operator.
-        ports.clear();
-        inv_ports.clear();
-    }
+  logger_->error(RSZ, 91, "unrecognized expr op from OpenSTA");
 }
 
-void
-RepairSetup::getEquivPortList(sta::FuncExpr *expr, sta::LibertyPortSet &ports)
+std::vector<bool> RepairSetup::simulateExpr(
+    sta::FuncExpr* expr,
+    sta::UnorderedMap<const LibertyPort*, std::vector<bool>>& port_stimulus)
+{
+  size_t table_length = 0x1 << port_stimulus.size();
+  std::vector<bool> result;
+  result.resize(table_length);
+  for (size_t i = 0; i < table_length; i++) {
+    result[i] = simulateExpr(expr, port_stimulus, i);
+  }
+
+  return result;
+}
+
+bool RepairSetup::isPortEqiv(sta::FuncExpr* expr,
+                             const LibertyCell* cell,
+                             const LibertyPort* port_a,
+                             const LibertyPort* port_b)
 {
-    sta::FuncExpr::Operator status = sta::FuncExpr::op_zero;
-    ports.clear();
-    sta::LibertyPortSet inv_ports;
-    getEquivPortList2(expr, ports, inv_ports, status);
-    if (inv_ports.size() > ports.size()) {
-        ports = inv_ports;
+  sta::LibertyCellPortIterator port_iter(cell);
+  sta::UnorderedMap<const LibertyPort*, std::vector<bool>> port_stimulus;
+  size_t input_port_count = 0;
+  while (port_iter.hasNext()) {
+    LibertyPort* port = port_iter.next();
+    if (port->direction()->isInput()) {
+      ++input_port_count;
+      port_stimulus[port] = {};
     }
-    if (status == sta::FuncExpr::op_port || ports.size() == 1) {
-        ports.clear();
+  }
+
+  if (input_port_count > 16) {
+    // Not worth manually simulating all these values.
+    // Probably need to do SAT solving or something else instead.
+    return false;
+  }
+
+  // Generate stimulus for the ports
+  size_t var_index = 0;
+  for (auto& it : port_stimulus) {
+    size_t truth_table_length = 0x1 << input_port_count;
+    std::vector<bool>& variable_stimulus = it.second;
+    variable_stimulus.resize(truth_table_length, false);
+    for (int i = 0; i < truth_table_length; i++) {
+      variable_stimulus[i] = static_cast<bool>((i >> var_index) & 0x1);
     }
+    var_index++;
+  }
+
+  if (port_stimulus.find(port_a) == port_stimulus.end()) {
+    return false;
+  }
+  if (port_stimulus.find(port_b) == port_stimulus.end()) {
+    return false;
+  }
```

Comment:
when would this happen?

Reply:
If someone used this function with a port that didn't belong to the cell.

Reply:
In that case I would just check port_a->libertyCell() == cell (likewise port_b) at the top of the method.